regex/src/main/scala/regex/Regex.scala

```diff
@@ -15,13 +29,51 @@ package regex
   */
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = 
+  lang match
+    case Concat(lang1, lang2) => 
+      if lang1 == Epsilon then simplify(lang2)
+      else if lang2 == Epsilon then simplify(lang1)
+      else if (lang1  == Empty || lang2 == Empty) then Empty
+      else Concat(simplify(lang1), simplify(lang2))
+
+    case Union(lang1, lang2) => 
+      if lang1 == Empty then simplify(lang2)
+      else if lang2 == Empty then simplify(lang1)
+      else Union(simplify(lang1), simplify(lang2))
+
+    case Star(x) => 
+      x match
+        case Epsilon => Epsilon
+        case Empty => Empty
+        case y => Star(simplify(y))
+      
+    case otherwise => otherwise
 
 /** A language is nullable if it contains ε */
-def nullable(lang: RegularLanguage): Boolean = ???
+def nullable(lang: RegularLanguage): Boolean = 
+  lang match
+    case Epsilon => true
+    case Empty => false
+    case Character(c) => false
+    case Union(lang1, lang2) => (nullable(lang1) || nullable(lang2))
+    case Concat(lang1, lang2) => (nullable(lang1) && nullable(lang2))
+    case Star(lang) => true
```

Comment:
One thing you could have done here is catch all the trues or falses with a wildcard at the end: case _ => true/false.  It would have saved you a line or two of code, but it's definitely more readable to list them all out like this